File: assignments/assignment-2-solution/src-csharp/TeaShop/UserInterface/PaymentBuilder/IPaymentBuilder.cs
using Assignment2Solution.Domain.Payment;

namespace Assignment2Solution.UserInterface.PaymentBuilder;

/// <summary>
///     Defines a payment method that can create a corresponding payment strategy.
/// </summary>
public interface IPaymentBuilder
{
    /// <summary>
    ///     Gets the display name of the payment method.
    /// </summary>
    string Name { get; }

    /// <summary>
    ///     Creates a payment strategy by prompting the user for necessary information.
    /// </summary>
    /// <param name="input">The text reader for user input.</param>
    /// <param name="output">The text writer for application output.</param>
    /// <returns>A new payment strategy.</returns>
    IPaymentStrategy CreateStrategy(TextReader input, TextWriter output);
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/userinterface/paymentbuilder/IPaymentBuilder.java
package assignment2solution.userinterface.paymentbuilder;

import assignment2solution.domain.payment.IPaymentStrategy;

import java.io.Reader;
import java.io.Writer;

/**
 * Defines a payment method that can create a corresponding payment strategy.
 */
public interface IPaymentBuilder {
    /**
     * Gets the display name of the payment method.
     */
    String getName();

    /**
     * Creates a payment strategy by prompting the user for necessary information.
     */
    IPaymentStrategy createStrategy(Reader input, Writer output);
}
